plugins/modules/foreman_setting.py

```diff
@@ -69,53 +67,45 @@
   description: Created / Updated state of the setting
 '''
 
-try:
-    from ansible.module_utils.ansible_nailgun_cement import (
-        find_setting,
-        ForemanAnsibleModule,
-        naildown_entity,
-        sanitize_entity_dict,
-    )
 
-    from nailgun.entities import (
-        Setting,
-    )
-except ImportError:
-    pass
+from ansible.module_utils.foreman_helper import ForemanApypieAnsibleModule, parameter_value_to_str
 
 
-name_map = {
-    'name': 'name',
-    'value': 'value',
+entity_spec = {
+    'name': {},
+    'value': {},
 }
 
 
 def main():
-    module = ForemanAnsibleModule(
+    module = ForemanApypieAnsibleModule(
         argument_spec=dict(
             name=dict(required=True),
-            value=dict(),
+            value=dict(type='raw'),
         ),
     )
 
-    entity_dict = module.parse_params()
+    entity_dict = module.clean_params()
 
     module.connect()
 
-    entity = find_setting(
-        module,
-        name=entity_dict['name'],
-        failsafe=False,
-    )
+    entity = module.find_resource_by_name('settings', entity_dict['name'], failsafe=False)
 
     if 'value' not in entity_dict:
-        entity_dict['value'] = entity.default or ""
+        entity_dict['value'] = entity['default'] or ''
+
+    settings_type = entity['settings_type']
+    new_value = entity_dict['value']
+    entity_dict['value'] = parameter_value_to_str(new_value, settings_type)
+    entity['value'] = parameter_value_to_str(entity['value'], settings_type)
 
-    entity_dict = sanitize_entity_dict(entity_dict, name_map)
+    changed, entity = module.ensure_entity('settings', entity_dict, entity, state='present', entity_spec=entity_spec)
 
-    changed, entity = naildown_entity(Setting, entity_dict, entity, 'present', module, check_type=True)
+    if entity:
+        # Fake the not serialized version as output
+        entity['value'] = new_value
```

Comment:
Why? When I `PUT` a new integer value (regardless whether the JSON I upload has it as a String or an Integer) the returned reply is an Integer, so it should compare correctly already?

```
# curl -X PUT -u admin:changeme -d '{"setting":{"value":361}}' -H 'Content-Type: application/json' -k https://foreman.example.com/api/v2/settings/132
{"value":361,"description":"Time in minutes installation tokens should be valid for, 0 to disable token generation","category":"Setting::Provisioning","settings_type":"integer","default":360,"created_at":"2019-07-16 06:55:35 UTC","updated_at":"2019-08-09 07:58:05 UTC","id":132,"name":"token_duration","full_name":"Token duration","category_name":"Provisioning"}

# curl -X PUT -u admin:changeme -d '{"setting":{"value":"362"}}' -H 'Content-Type: application/json' -k https://foreman.example.com/api/v2/settings/132
{"value":362,"description":"Time in minutes installation tokens should be valid for, 0 to disable token generation","category":"Setting::Provisioning","settings_type":"integer","default":360,"created_at":"2019-07-16 06:55:35 UTC","updated_at":"2019-08-09 07:59:13 UTC","id":132,"name":"token_duration","full_name":"Token duration","category_name":"Provisioning"}
```

Reply:
The problem is not with integers:
`"msg": "Setting the setting failed! ('[\"foo\", \"bar\"]' != '[u'foo', u'bar']')"`

Reply:
Right. But it's the integers where it bites us in the tests. So my thought to update the `if` to:

```python
if entity and entity['settings_type'] in ['array', 'hash', 'yaml', 'json']:
  entity['value'] = new_value
```

Reply:
Hmm. It's more complicated than this, because the api wants `string` and gives us `integer`.
So in case, we did not update the setting, the only reasonable thing to return is the "unchanged" current entity. However we changed it already, to make it fit for the comparison in `_update_entity`. So the solution, i found, is to convert the input to integer, if the `settings_type` is `integer`.

Reply:
Works for me, thanks!